snap/quota/quota.go

```diff
@@ -182,22 +251,76 @@ func (grp *Group) validate() error {
 		}
 	}
 
-	// check that if this is a sub-group, then the parent group has enough space
+	// Check that if this is a sub-group, then the parent group has enough space
 	// to accommodate this new group (we assume that other existing sub-groups
-	// in the parent group have already been validated)
+	// in the parent group have already been validated).
 	if grp.parentGroup != nil {
-		alreadyUsed := quantity.Size(0)
+		memoryUsed := quantity.Size(0)
+		tasksUsed := 0
+		cpuQuotaUsed := 0
+
+		max := func(a, b int) int {
+			if a > b {
+				return a
+			}
+			return b
+		}
+
+		calculateCpuQuota := func(grp *Group) int {
+			if grp.CpuLimit == nil {
+				return 0
+			}
+			return max(grp.CpuLimit.Count, 1) * grp.CpuLimit.Percentage
+		}
+
 		for _, child := range grp.parentGroup.subGroups {
 			if child.Name == grp.Name {
 				continue
 			}
-			alreadyUsed += child.MemoryLimit
+
+			memoryUsed += child.MemoryLimit
+			tasksUsed += child.TaskLimit
+			cpuQuotaUsed += calculateCpuQuota(child)
+		}
+
+		if grp.parentGroup.CpuLimit != nil && grp.CpuLimit != nil {
+			arrayContains := func(arr []int, val int) bool {
+				for _, v := range arr {
+					if v == val {
+						return true
+					}
+				}
+				return false
+			}
+
+			upperLimit := calculateCpuQuota(grp.parentGroup)
+			if cpuQuotaUsed+calculateCpuQuota(grp) > upperLimit {
+				return fmt.Errorf("group %q would exceed its parent group's CPU limit", grp.Name)
+			}
+
+			if len(grp.parentGroup.CpuLimit.AllowedCPUs) > 0 {
```

Comment:
do we need to consider the nested nested case where foo1 is the parent group of foo2 and foo2 is the parent group of foo3 which is a real group and has snaps in it, and only foo1 has the AllowedCPUs set but foo2 does not specify any cpu resources but instead specifies memory limits? 

also I guess more generally are we doing enough validation to ensure that when creating sub-groups that resources specified by the parent of a parent are considered when the first parent does not specify that resource? It almost seems like we need some tree traversal functions here where we find the first parent of our current group that has the same resource type we are considering and treat that as the "real parent limit"

Reply:
I have created a tree travel function now that (hopefully) correctly validates these scenarios. I have also added a few tests to verify it catches quotas across sub-groups. I might have missed a few interesting cases in the tests. But this was a really awesome catch, and was a weakness I did not think off. Thanks for your review so far!